### PR TITLE
[feat][client] New version diskcache metrics collection

### PR DESCRIPTION
**Pull request by meilirensheng2020**

### CODE REVIEW

File: curvefs/src/client/blockcache/block_cache.cpp [hunk 1]
```diff
@@ -104,7 +108,12 @@ BCACHE_ERROR BlockCacheImpl::Put(const BlockKey& key, const Block& block) {
 
   // TODO: Cache the block which put to storage directly
   timer.NextPhase(Phase::S3_PUT);
-  return s3_->Put(key.StoreKey(), block.data, block.size);
+  rc = s3_->Put(key.StoreKey(), block.data, block.size);
+  if (rc == BCACHE_ERROR::OK) {
+    CollectMetrics(&s3Metric_->adaptorWriteS3, block.size,
+                   butil::cpuwide_time_us() - start);
```
Comment (Wine93) [inline curvefs/src/client/blockcache/block_cache.cpp:114]: Maybe you should use the other timer for `adaptorWriteS3`.

File: curvefs/src/client/blockcache/block_cache.cpp [hunk 2]
```diff
@@ -196,6 +219,10 @@ void BlockCacheImpl::UploadStageBlock(const BlockKey& key,
       LOG(ERROR) << "Remove stage (" << key.Filename()
                  << ") failed: " << StrErr(rc);
     }
+    if (rc == BCACHE_ERROR::OK) {
+      CollectMetrics(&s3Metric_->adaptorWriteS3, length,
+                     butil::cpuwide_time_us() - start);
```
Comment (Wine93) [inline curvefs/src/client/blockcache/block_cache.cpp:224]: ditto.

File: curvefs/src/client/blockcache/block_cache.cpp [hunk 3]
```diff
@@ -40,6 +40,7 @@ namespace curvefs {
 namespace client {
 namespace blockcache {
 
+using ::curvefs::client::metric::CollectMetrics;
```
Comment (Wine93) [inline curvefs/src/client/blockcache/block_cache.cpp:43]: Add a newline here.

**Comment by Wine93** [comment]
I think maybe use two alone metrics is better, the one for `disk_cache` and the another for `s3_client`, also i think it's easy to implement.

File: curvefs/src/client/blockcache/block_cache.h [hunk 4]
```diff
@@ -72,6 +74,8 @@ class BlockCache {
   virtual bool IsCached(const BlockKey& key) = 0;
 
   virtual StoreType GetStoreType() = 0;
+
+  virtual void InitS3Metrics(std::shared_ptr<S3Metric> s3Metric_) = 0;
```
[the remark this answers is not in the document]
Comment (meilirensheng2020) [reply curvefs/src/client/blockcache/block_cache.h]: metrics collection move to s3_clent

File: curvefs/src/client/blockcache/block_cache.h [hunk 5]
```diff
@@ -96,6 +100,9 @@ class BlockCacheImpl : public BlockCache {
   bool IsCached(const BlockKey& key) override;
 
   StoreType GetStoreType() override;
+  void InitS3Metrics(std::shared_ptr<S3Metric> s3Metric_) {
+    this->s3Metric_ = s3Metric_;
```
[the remark this answers is not in the document]
Comment (meilirensheng2020) [reply curvefs/src/client/blockcache/block_cache.h]: ditto

File: curvefs/src/client/blockcache/block_cache.h [hunk 6]
```diff
@@ -96,6 +100,9 @@ class BlockCacheImpl : public BlockCache {
   bool IsCached(const BlockKey& key) override;
 
   StoreType GetStoreType() override;
```
[the remark this answers is not in the document]
Comment (meilirensheng2020) [reply curvefs/src/client/blockcache/block_cache.h]: ditto

File: curvefs/src/client/blockcache/block_cache.cpp (hunk 1 above)
Comment (meilirensheng2020) [reply curvefs/src/client/blockcache/block_cache.cpp]: done.

File: curvefs/src/client/blockcache/block_cache.cpp (hunk 2 above)
Comment (meilirensheng2020) [reply curvefs/src/client/blockcache/block_cache.cpp]: done.

File: curvefs/src/client/blockcache/disk_cache.cpp [hunk 7]
```diff
@@ -42,13 +43,19 @@ namespace blockcache {
 using ::curvefs::base::string::GenUuid;
 using ::curvefs::base::time::TimeNow;
 
+using DiskCacheTotalMetric = ::curvefs::client::metric::DiskCacheMetric;
+using DiskCacheMetricGuard =
+    ::curvefs::client::blockcache::DiskCacheMetricGuard;
```
Comment (Wine93) [inline curvefs/src/client/blockcache/disk_cache.cpp:48]: Add a newline here.

File: curvefs/src/client/blockcache/disk_cache_metric.h [hunk 8]
```diff
@@ -135,13 +137,35 @@ using MetricHandler = std::function<void(DiskCacheMetric* metric)>;
...
 
   std::shared_ptr<DiskCacheMetric> metric;
   MetricHandler handler;
 };
+struct DiskCacheMetricGuard {
+  explicit DiskCacheMetricGuard(BCACHE_ERROR* rc, InterfaceMetric* metric,
+                                size_t count)
+      : rc_(rc), metric_(metric), count_(count) {
+    start_ = butil::cpuwide_time_us();
+  }
```
Comment (Wine93) [inline curvefs/src/client/blockcache/disk_cache_metric.h:152]: ditto.

File: curvefs/src/client/blockcache/disk_cache_metric.h [hunk 9]
```diff
@@ -135,13 +137,35 @@ using MetricHandler = std::function<void(DiskCacheMetric* metric)>;
...
+    if (*rc_ == BCACHE_ERROR::OK) {
+      metric_->bps.count << count_;
+      metric_->qps.count << 1;
+      auto duration = butil::cpuwide_time_us() - start_;
+      metric_->latency << duration;
+      metric_->latTotal << duration;
+    } else {
+      metric_->eps.count << 1;
+    }
+  }
```
Comment (Wine93) [inline curvefs/src/client/blockcache/disk_cache_metric.h:163]: ditto.

File: curvefs/src/client/blockcache/s3_client.cpp [hunk 10]
```diff
@@ -68,6 +83,9 @@ void S3ClientImpl::AsyncPut(const std::string& key, const char* buffer,
   context->startTime = butil::cpuwide_time_us();
   context->cb = [&,
                  retry](const std::shared_ptr<PutObjectAsyncContext>& context) {
+    ;
```
Comment (Wine93) [inline curvefs/src/client/blockcache/s3_client.cpp:86]: `;` is redundant.

File: curvefs/src/client/metric/client_metric.cpp [hunk 11]
```diff
@@ -40,6 +42,10 @@ const std::string KVClientMetric::prefix = "dingofs_kvclient";        // NOLINT
 const std::string S3ChunkInfoMetric::prefix = "inode_s3_chunk_info";  // NOLINT
 const std::string WarmupManagerS3Metric::prefix = "dingofs_warmup";   // NOLINT
 
+// inititalize mutex (Singleton mode)
+std::mutex DiskCacheMetric::mtx_;
```
Comment (Wine93) [inline curvefs/src/client/metric/client_metric.cpp:46]: It's confusing.

File: curvefs/src/client/blockcache/disk_cache.cpp (hunk 7 above)
Comment (meilirensheng2020) [reply curvefs/src/client/blockcache/disk_cache.cpp:48]: done

File: curvefs/src/client/blockcache/disk_cache_metric.h (hunk 8 above)
Comment (meilirensheng2020) [reply curvefs/src/client/blockcache/disk_cache_metric.h:152]: done

File: curvefs/src/client/blockcache/disk_cache_metric.h (hunk 9 above)
Comment (meilirensheng2020) [reply curvefs/src/client/blockcache/disk_cache_metric.h:164]: done

File: curvefs/src/client/metric/client_metric.cpp (hunk 11 above)
Comment (meilirensheng2020) [reply curvefs/src/client/metric/client_metric.cpp]: This is the initialization of static variables（static std::mutex mtx_；）, which is no longer necessary at present。
already remove。

File: curvefs/src/client/blockcache/s3_client.cpp (hunk 10 above)
Comment (meilirensheng2020) [reply curvefs/src/client/blockcache/s3_client.cpp]: done。

**Review by Wine93 — APPROVED** [review]

**Pull request merged**